pkg/kubelet/cm/memorymanager/policy_single_numa.go

```diff
@@ -92,45 +103,81 @@ func (p *singleNUMAPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Co
 	hint := p.affinity.GetAffinity(string(pod.UID), container.Name)
 	klog.Infof("[memorymanager] Pod %v, Container %v Topology Affinity is: %v", pod.UID, container.Name, hint)
 
-	var affinityBits []int
-	if !hint.Preferred || hint.NUMANodeAffinity == nil {
-		defaultAffinity, err := getDefaultNUMAAffinity(s, container)
+	requestedResources, err := getRequestedResources(container)
+	if err != nil {
+		return err
+	}
+
+	bestHint := &hint
+	// topology manager returned the hint with NUMA affinity nil
+	// we should use the default NUMA affinity calculated the same way as for the topology manager
+	if hint.NUMANodeAffinity == nil {
+		defaultHint, err := p.getDefaultHint(s, requestedResources)
 		if err != nil {
 			return err
 		}
-		affinityBits = defaultAffinity.GetBits()
-	} else {
-		affinityBits = hint.NUMANodeAffinity.GetBits()
-	}
 
-	if len(affinityBits) > 1 {
-		return fmt.Errorf("[memorymanager] NUMA affinity has more than one NUMA node")
+		if !defaultHint.Preferred && bestHint.Preferred {
+			return fmt.Errorf("[memorymanager] failed to find the default preferred hint")
+		}
+		bestHint = defaultHint
 	}
 
 	machineState := s.GetMachineState()
 
-	var containerBlocks []state.Block
-	for resourceName, q := range container.Resources.Requests {
-		if resourceName != v1.ResourceMemory && !corehelper.IsHugePageResourceName(resourceName) {
-			continue
+	// topology manager returns the hint that does not satisfy completely the container request
+	// we should extend this hint to the one who will satisfy the request and include the current hint
+	if !isAffinitySatisfyRequest(machineState, bestHint.NUMANodeAffinity, requestedResources) {
+		extendedHint, err := p.extendTopologyManagerHint(s, requestedResources, bestHint.NUMANodeAffinity)
+		if err != nil {
+			return err
 		}
 
-		size, succeed := q.AsInt64()
-		if !succeed {
-			return fmt.Errorf("[memorymanager] failed to represent quantity as int64")
+		if !extendedHint.Preferred && bestHint.Preferred {
+			return fmt.Errorf("[memorymanager] failed to find the extended preferred hint")
 		}
+		bestHint = extendedHint
+	}
```

Comment:
As I understand extending hints is needed when **"restricted"** policy of Topology Manager is used. In case of other policies, i.e. **"single-numa-node"** this won't be used as TM will always return a hint that satisfies completely the container request or he rejects the pod admission. 

If that's right, I think it's hard to know the purpose of `extendTopologyManagerHint` only based on the code so there may be a good place for a broader explanation of that.

Reply:
Added the proper description on top of `extendTopologyManagerHint`.